File: 그래프_bfs, 최단거리문제/leetcode994(Medium)-Rotting Oranges.js
// BFS : 75ms, 44.5MB
/**
 * @param {number[][]} grid
 * @return {number}
 */
var orangesRotting = function (grid) {
  const N = grid.length
  const M = grid[0].length

  const dx = [0, 1, 0, -1]
  const dy = [1, 0, -1, 0]

  const validateRange = (nx, ny) => {
    return 0 <= nx && nx < N && 0 <= ny && ny < M
  }

  const rottens = []
  let entireCount = 0
  for (let i = 0; i < N; i += 1) {
    for (let j = 0; j < M; j += 1) {
      if (grid[i][j] !== 0) {
        entireCount += 1
        if (grid[i][j] === 2) {
          rottens.push([i, j, 0])
        }
      }
    }
  }

  let rottenCount = rottens.length
  let result = 0
  while (rottens.length > 0) {
    const [x, y, time] = rottens.shift()
    result = Math.max(result, time)
    for (let i = 0; i < 4; i += 1) {
      const nx = x + dx[i]
      const ny = y + dy[i]
      if (validateRange(nx, ny) && grid[nx][ny] === 1) {
        rottenCount += 1
        grid[nx][ny] = 2
        rottens.push([nx, ny, time + 1])
      }
    }
  }

  return entireCount - rottenCount > 0 ? -1 : result
}

File: 그래프_bfs, 최단거리문제/leetcode994(Medium)-Rotting Oranges.py
# BFS : 53ms, 13.9MB
class Solution:
    def orangesRotting(self, grid: List[List[int]]) -> int:
        N = len(grid)
        M = len(grid[0])
        
        dx = [-1, 0, 1, 0]
        dy = [0, -1, 0, 1]
        
        def validate_range(nx, ny):
            return 0 <= nx < N and 0 <= ny < M
        
        entire_count = 0
        rottens = collections.deque()
        for x in range(N):
            for y in range(M):
                if grid[x][y] != 0:
                    entire_count += 1
                    if grid[x][y] == 2:
                        rottens.append((x, y, 0))
        
        rotten_count = len(rottens)
        
        result = 0
        while rottens:
            x, y, time = rottens.popleft()
            result = max(result, time)
            for i in range(4):
                nx = x + dx[i]
                ny = y + dy[i]
                if validate_range(nx, ny) and grid[nx][ny] == 1:
                    rotten_count += 1
                    grid[nx][ny] = 2
                    rottens.append((nx, ny, time + 1))
            
            
        if entire_count - rotten_count > 0 :
            return -1
        return result